chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,325 @@
+    ra = None
+    dec = None
+    centroid_dt = None
+    _dt_applied = None
+
+    def __init__(self, start, stop):
+        self.start = start
+        self.stop = stop
+
+
+    def set_centroids(self, source, slot, alg=8, apply_dt=True):
+        """
+        Get centroids from ``source``.
+
+        One could also just set yags, yag_times, zags, zag_times
+        attributes directly.
+
+        :param source: 'ground' | 'obc'
+        :param slot: ACA slot
+        """
+        self.centroid_source = source
+        start = self.start
+        stop = self.stop
+        # Get centroids from Ska eng archive or mica L1 archive
+        # Might want to add filtering for track status here too
+        # Also need to include something for time offsets
+        # And warn to console if we hit multiple obsids
+        if source == 'ground':
+            acen_files = sorted(asp_l1.get_files(start=start, stop=stop, content=['ACACENT']))
+            acen = vstack([Table.read(f) for f in sorted(acen_files)], metadata_conflicts='silent')
+            ok = ((acen['slot'] == slot) & (acen['alg'] == alg) & (acen['status'] == 0)
+                  & (acen['time'] >= DateTime(start).secs) & (acen['time'] <= DateTime(stop).secs))
+            yags = np.array(acen[ok]['ang_y'] * 3600)
+            zags = np.array(acen[ok]['ang_z'] * 3600)
+            yag_times = np.array(acen[ok]['time'])
+            zag_times = np.array(acen[ok]['time'])
+        elif source == 'obc':
+            telem = fetch.Msidset(['AOACYAN{}'.format(slot), 'AOACZAN{}'.format(slot)], start, stop)
+            # Filter centroids for reasonble-ness
+            yok = telem['AOACYAN{}'.format(slot)].vals != -3276.8
+            zok = telem['AOACZAN{}'.format(slot)].vals != -3276.8
+            yags = telem['AOACYAN{}'.format(slot)].vals[yok]
+            yag_times = telem['AOACYAN{}'.format(slot)].times[yok]
+            zags = telem['AOACZAN{}'.format(slot)].vals[zok]
+            zag_times = telem['AOACZAN{}'.format(slot)].times[zok]
+        else:
+            raise ValueError("centroid_source must be 'obc' or 'ground'")
+        self.yags = yags
+        self.yag_times = yag_times
+        self.zags = zags
+        self.zag_times = zag_times
+        if apply_dt:
+            self.set_offsets()
+
+
+    def set_atts(self, source):
+        """Get attitude solution quaternions from ``source``.
+
+        One could also just set atts and att_times attributes directly.
+        """
+        self.att_source = source
+        tstart = DateTime(self.start).secs
+        tstop = DateTime(self.stop).secs
+        # Get attitudes and times
+        if source == 'obc':
+            telem = fetch.Msidset(['aoattqt*'], tstart, tstop)
+            atts = np.vstack([telem['aoattqt{}'.format(idx)].vals
+                              for idx in [1, 2, 3, 4]]).transpose()
+            att_times = telem['aoattqt1'].times
+            obsids = fetch.Msid('COBSRQID', tstart, tstop)
```

Comment:
I suppose it would be both safer and still somewhat protect to fetch a few seconds at the beginning *and* and of the interval.  That would protect the code for weird results if the range covers multiple obsids.